File: SBSAnimoji/Pages/Main/MainView.m
//
//  MainView.m
//  SBSAnimoji
//
//  Created by Simon Støvring on 05/11/2017.
//  Copyright © 2017 SimonBS. All rights reserved.
//

#import "MainView.h"
#import "AVTPuppetView.h"

@interface MainView ()
@property (nonatomic, strong) SBSPuppetView *puppetView;
@property (nonatomic, strong) UICollectionView *thumbnailsCollectionView;
@property (nonatomic, strong) UIButton *recordButton;
@property (nonatomic, strong) UIButton *shareButton;
@property (nonatomic, strong) UIButton *deleteButton;
@property (nonatomic, strong) UIButton *previewButton;
@property (nonatomic, strong) UIActivityIndicatorView *activityIndicatorView;
@property (nonatomic, strong) UILabel *durationLabel;
@property (nonatomic, strong) UIView *puppetViewSeparatorView;
@end

@implementation MainView

- (instancetype)init {
    if (self = [super init]) {
        [self setupView];
        [self setupLayout];
    }
    return self;
}

- (void)setupView {
    self.backgroundColor = [UIColor whiteColor];
    
    self.puppetView = [[SBSPuppetView alloc] init];
    self.puppetView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.puppetView];
    
    self.puppetViewSeparatorView = [[UIView alloc] init];
    self.puppetViewSeparatorView.translatesAutoresizingMaskIntoConstraints = NO;
    self.puppetViewSeparatorView.backgroundColor = [UIColor colorWithRed:224.0/255.0 green:224.0/255.0 blue:224.0/255.0 alpha:1.0];
    [self addSubview:self.puppetViewSeparatorView];
    
    UICollectionViewFlowLayout *collectionViewLayout = [[UICollectionViewFlowLayout alloc] init];
    collectionViewLayout.itemSize = CGSizeMake(80, 80);
    collectionViewLayout.scrollDirection = UICollectionViewScrollDirectionVertical;
    collectionViewLayout.minimumInteritemSpacing = 14;
    collectionViewLayout.minimumLineSpacing = 10;
    self.thumbnailsCollectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:collectionViewLayout];
    self.thumbnailsCollectionView.translatesAutoresizingMaskIntoConstraints = NO;
    self.thumbnailsCollectionView.backgroundColor = [UIColor whiteColor];
    self.thumbnailsCollectionView.contentInset = UIEdgeInsetsMake(15, 7, 15, 7);
    self.thumbnailsCollectionView.showsHorizontalScrollIndicator = NO;
    [self addSubview:self.thumbnailsCollectionView];
    
    self.durationLabel = [[UILabel alloc] init];
    self.durationLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.durationLabel.textAlignment = NSTextAlignmentRight;
    self.durationLabel.font = [UIFont systemFontOfSize:14 weight:UIFontWeightMedium];
    self.durationLabel.hidden = YES;
    [self addSubview:self.durationLabel];
    
    self.deleteButton = [UIButton buttonWithType:UIButtonTypeSystem];
    self.deleteButton.translatesAutoresizingMaskIntoConstraints = NO;
    self.deleteButton.hidden = YES;
    [self.deleteButton setImage:[UIImage imageNamed:@"delete"] forState:UIControlStateNormal];
    [self addSubview:self.deleteButton];
    
    self.previewButton = [UIButton buttonWithType:UIButtonTypeSystem];
    self.previewButton.translatesAutoresizingMaskIntoConstraints = NO;
    self.previewButton.hidden = YES;
    [self.previewButton setImage:[UIImage imageNamed:@"play"] forState:UIControlStateNormal];
    [self addSubview:self.previewButton];
    
    self.recordButton = [UIButton buttonWithType:UIButtonTypeSystem];
    self.recordButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.recordButton setImage:[UIImage imageNamed:@"record"] forState:UIControlStateNormal];
    [self addSubview:self.recordButton];
    
    self.shareButton = [UIButton buttonWithType:UIButtonTypeSystem];
    self.shareButton.translatesAutoresizingMaskIntoConstraints = NO;
    self.shareButton.hidden = YES;
    [self.shareButton setImage:[UIImage imageNamed:@"share"] forState:UIControlStateNormal];
    [self addSubview:self.shareButton];
    
    self.activityIndicatorView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    self.activityIndicatorView.translatesAutoresizingMaskIntoConstraints = NO;
    self.activityIndicatorView.hidesWhenStopped = YES;
    [self addSubview:self.activityIndicatorView];
}

- (void)setupLayout {
    [self.puppetView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor].active = YES;
    [self.puppetView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor].active = YES;
    [self.puppetView.topAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.topAnchor].active = YES;
    [self.puppetView.heightAnchor constraintEqualToConstant:335].active = YES;
    
    [self.puppetViewSeparatorView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor].active = YES;
    [self.puppetViewSeparatorView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor].active = YES;
    [self.puppetViewSeparatorView.topAnchor constraintEqualToAnchor:self.puppetView.bottomAnchor].active = YES;
    [self.puppetViewSeparatorView.heightAnchor constraintEqualToConstant:2].active = YES;
    
    [self.thumbnailsCollectionView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor].active = YES;
    [self.thumbnailsCollectionView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor].active = YES;
    [self.thumbnailsCollectionView.topAnchor constraintEqualToAnchor:self.puppetViewSeparatorView.bottomAnchor].active = YES;
    [self.thumbnailsCollectionView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor].active = YES;
    
    [self.durationLabel.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant: -20].active = YES;
    [self.durationLabel.topAnchor constraintEqualToAnchor:self.puppetView.topAnchor constant: 15].active = YES;
    
    [self.recordButton.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant: -20].active = YES;
    [self.recordButton.bottomAnchor constraintEqualToAnchor:self.puppetView.bottomAnchor constant: -20].active = YES;

    [self.shareButton.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant: -20].active = YES;
    [self.shareButton.bottomAnchor constraintEqualToAnchor:self.puppetView.bottomAnchor constant: -20].active = YES;

    [self.deleteButton.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant: -20].active = YES;
    [self.deleteButton.topAnchor constraintEqualToAnchor:self.puppetView.topAnchor constant: 15].active = YES;
    
    [self.previewButton.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant: -20].active = YES;
    [self.previewButton.topAnchor constraintEqualToAnchor:self.deleteButton.bottomAnchor constant: 15].active = YES;

    [self.activityIndicatorView.centerXAnchor constraintEqualToAnchor:self.shareButton.centerXAnchor].active = YES;
    [self.activityIndicatorView.centerYAnchor constraintEqualToAnchor:self.shareButton.centerYAnchor].active = YES;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    CGFloat itemsPerRow = 4;
    UICollectionView *collectionView = self.thumbnailsCollectionView;
    UIEdgeInsets contentInset = self.thumbnailsCollectionView.contentInset;
    UICollectionViewFlowLayout *flowLayout = (UICollectionViewFlowLayout *)collectionView.collectionViewLayout;
    CGFloat availableWidth = self.bounds.size.width - contentInset.left - contentInset.right - (itemsPerRow - 1) * flowLayout.minimumInteritemSpacing;
    CGFloat itemLength = floor(availableWidth / itemsPerRow);
    flowLayout.itemSize = CGSizeMake(itemLength, itemLength);
}

@end

File: Animoji/Model/MainView.swift
//
//  MainView.swift
//  Animoji
//
//  Created by Daniel Illescas Romero on 25/12/2017.
//  Copyright © 2017 Daniel Illescas Romero. All rights reserved.
//

import UIKit

@objc class MainView: UIView {
	
	var puppetView = SBSPuppetView()
	var thumbnailsCollectionView = UICollectionView(frame: .zero, collectionViewLayout: UICollectionViewLayout())
	var recordButton = UIButton(type: .system)
	var shareButton = UIButton(type: .system)
	var deleteButton = UIButton(type: .system)
	var previewButton = UIButton(type: .system)
	var activityIndicatorView = UIActivityIndicatorView()
	var durationLabel = UILabel()
	var puppetViewSeparatorView = UIView()
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		self.setupView()
		self.setupLayout()
	}
	
	func setupView() {

		self.backgroundColor = .white
		
		self.puppetView.translatesAutoresizingMaskIntoConstraints = false
		self.addSubview(self.puppetView)
		
		self.puppetViewSeparatorView.translatesAutoresizingMaskIntoConstraints = false
		self.puppetViewSeparatorView.backgroundColor = UIColor(red: 224.0/255.0, green: 224.0/255.0, blue: 224.0/255.0, alpha: 1.0)
		self.addSubview(self.puppetViewSeparatorView)
		
		let collectionViewLayout = UICollectionViewFlowLayout()
		collectionViewLayout.scrollDirection = .vertical
		collectionViewLayout.minimumInteritemSpacing = 14
		collectionViewLayout.minimumLineSpacing = 10
		
		self.thumbnailsCollectionView.setCollectionViewLayout(collectionViewLayout, animated: true)
		self.thumbnailsCollectionView.translatesAutoresizingMaskIntoConstraints = false
		self.thumbnailsCollectionView.backgroundColor = .white
		self.thumbnailsCollectionView.contentInset = UIEdgeInsets(top: 15, left: 7, bottom: 15, right: 7)
		self.thumbnailsCollectionView.showsHorizontalScrollIndicator = false
		self.addSubview(self.thumbnailsCollectionView)
		
		self.durationLabel.translatesAutoresizingMaskIntoConstraints = false
		self.durationLabel.textAlignment = .right
		self.durationLabel.font = .systemFont(ofSize: 14, weight: .medium)
		self.durationLabel.isHidden = true
		self.addSubview(self.durationLabel)
		
		self.deleteButton.translatesAutoresizingMaskIntoConstraints = false
		self.deleteButton.isHidden = true
		self.deleteButton.setImage(#imageLiteral(resourceName: "delete"), for: .normal)
		self.addSubview(self.deleteButton)
		
		self.previewButton.translatesAutoresizingMaskIntoConstraints = false
		self.previewButton.isHidden = true
		self.previewButton.setImage(#imageLiteral(resourceName: "start-previewing"), for: .normal)
		self.addSubview(self.previewButton)
		
		self.recordButton.translatesAutoresizingMaskIntoConstraints = false
		self.recordButton.setImage(#imageLiteral(resourceName: "start-recording"), for: .normal)
		self.addSubview(self.recordButton)
		
		self.shareButton.translatesAutoresizingMaskIntoConstraints = false
		self.shareButton.isHidden = true
		self.shareButton.setImage(#imageLiteral(resourceName: "share"), for: .normal)
		self.addSubview(self.shareButton)
		
		self.activityIndicatorView.translatesAutoresizingMaskIntoConstraints = false
		self.activityIndicatorView.hidesWhenStopped = true
		self.addSubview(self.activityIndicatorView)
	}
	
	func setupLayout() {
		
		self.puppetView.leadingAnchor.constraint(equalTo: self.leadingAnchor).isActive = true
		self.puppetView.trailingAnchor.constraint(equalTo: self.trailingAnchor).isActive = true
		self.puppetView.topAnchor.constraint(equalTo: self.safeAreaLayoutGuide.topAnchor).isActive = true
		self.puppetView.heightAnchor.constraint(equalToConstant: 355).isActive = true
		
		self.puppetViewSeparatorView.leadingAnchor.constraint(equalTo: self.leadingAnchor).isActive = true
		self.puppetViewSeparatorView.trailingAnchor.constraint(equalTo: self.trailingAnchor).isActive = true
		self.puppetViewSeparatorView.topAnchor.constraint(equalTo: self.puppetView.bottomAnchor).isActive = true
		self.puppetViewSeparatorView.heightAnchor.constraint(equalToConstant: 2).isActive = true
		
		self.thumbnailsCollectionView.leadingAnchor.constraint(equalTo: self.leadingAnchor).isActive = true
		self.thumbnailsCollectionView.trailingAnchor.constraint(equalTo: self.trailingAnchor).isActive = true
		self.thumbnailsCollectionView.topAnchor.constraint(equalTo: self.puppetViewSeparatorView.bottomAnchor).isActive = true
		self.thumbnailsCollectionView.bottomAnchor.constraint(equalTo: self.bottomAnchor).isActive = true
		
		self.durationLabel.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: -20).isActive = true
		self.durationLabel.topAnchor.constraint(equalTo: self.puppetView.topAnchor, constant: 15).isActive = true
		
		self.recordButton.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: -20).isActive = true
		self.recordButton.bottomAnchor.constraint(equalTo: self.puppetView.bottomAnchor, constant: -20).isActive = true
		
		self.shareButton.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: -20).isActive = true
		self.shareButton.bottomAnchor.constraint(equalTo: self.puppetView.bottomAnchor, constant: -20).isActive = true
		
		self.deleteButton.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: -20).isActive = true
		self.deleteButton.topAnchor.constraint(equalTo: self.puppetView.topAnchor, constant: 15).isActive = true
		
		self.previewButton.trailingAnchor.constraint(equalTo: self.trailingAnchor, constant: -20).isActive = true
		self.previewButton.topAnchor.constraint(equalTo: self.deleteButton.bottomAnchor, constant: 15).isActive = true
		
		self.activityIndicatorView.centerXAnchor.constraint(equalTo: self.shareButton.centerXAnchor).isActive = true
		self.activityIndicatorView.centerYAnchor.constraint(equalTo: self.shareButton.centerYAnchor).isActive = true
	}
	
	override func layoutSubviews() {
		super.layoutSubviews()
		
		let itemsPerRow: CGFloat = 4
		let collectionView = self.thumbnailsCollectionView
		let contentInset = collectionView.contentInset
		
		if let flowLayout = collectionView.collectionViewLayout as? UICollectionViewFlowLayout {
			let availableWidth = self.bounds.size.width - contentInset.left - contentInset.right - (itemsPerRow - 1) * flowLayout.minimumInteritemSpacing
			let itemLength = floor(availableWidth / itemsPerRow)
			flowLayout.itemSize = CGSize(width: itemLength, height: itemLength)
		}
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
}
